smol/moca/utils/math_utils.py

```diff
@@ -0,0 +1,934 @@
+    """
+
+    def test_connected(vs, ns_disconnected, ns):
+        return np.array([is_connected(n, vs, ns) for n in ns_disconnected], dtype=bool)
+
+    def candidate_key(u):
+        return flip_size(u)
+
+    n0 = np.array(n0, dtype=int)
+    xs = np.array(xs, dtype=int)
+    vs = np.array(vs, dtype=int)
+    ns = xs @ vs + n0
+    connected = test_connected(vs, ns, ns)
+    ns_disconnected = ns[~connected]
+    if len(ns_disconnected) == 0:
+        return vs
+
+    tree = KDTree(ns)
+    candidate_vectors = []
+    # Get k nearest neighbors in Euclidean distance.
+    for n in ns_disconnected:
+        dists, point_ids = tree.query(n, k=k)
+        if dists[0] == 0:
+            point_ids = point_ids[1:]
+        points = ns[point_ids, :]
+        for point in points:
+            u = point - n
+            if (
+                tuple(u.tolist()) in candidate_vectors
+                or tuple((-u).tolist()) in candidate_vectors
+            ):
+                continue
+            candidate_vectors.append(tuple(u.tolist()))
+
+    candidate_vectors = sorted(candidate_vectors, key=candidate_key)
+    candidate_vectors = np.array(candidate_vectors, dtype=int)
+    selected_vectors = vs.copy()
+    ns_rem = ns_disconnected.copy()
+    for u in candidate_vectors:
+        vs_current = np.concatenate((selected_vectors, [u]), axis=0)
+        connected = test_connected(vs_current, ns_rem, ns)
+        selected_vectors = vs_current.copy()
+        ns_rem = ns_rem[~connected]
+        if len(ns_rem) == 0:
+            break
+
+    return selected_vectors
+
+
+def flip_weights_mask(flip_vectors, n, max_n=None):
+    """Mark feasibility of flip vectors.
+
+    If a flip direction leads to any n+v < 0, then it is marked
+    infeasible. Generates a boolean mask, every two components
+    marks whether a flip direction and its inverse is feasible
+    given n at the current occupancy.
+    Will be used by Tableflipper.
+
+    Args:
+        flip_vectors(1D ArrayLike[int]):
+            Flip directions in the table (inverses not included).
+        n(1D ArrayLike[int]):
+            Amount of each specie on sublattices. Same as returned
+            by occu_to_species_n.
+        max_n(1D ArrayLike[int]): optional
+            Maximum number of each species allowed. This is needed
+            When the number of active sites != number of sites in
+            some sub-lattice.
+
+    Return:
+        Direction and its inverse are feasible or not:
+           1D np.ndarray[bool]
+    """
+    flip_vectors = np.array(flip_vectors, dtype=int)
+    directions = np.concatenate([(u, -u) for u in flip_vectors], axis=0)
+    if max_n is None:
+        max_n = np.ones(len(n)) * np.inf
+    elif isinstance(max_n, (int, np.int32, np.int64)):
+        max_n = np.ones(len(n), dtype=int) * max_n
+    else:
+        max_n = np.array(max_n, dtype=int)
+    return ~(
+        np.any(directions + n < 0, axis=-1) | np.any(directions + n > max_n, axis=-1)
+    )
```

Comment:
Finding ergodic and optimal vectors might only be used by CompositionSpace object, but I prefer not to put all these into comp_space.py because this can make the comp_space file too bulky. 

flip_weights_mask might be called by future modification of mckernel as well, for example, in wang-landau or metropolis with a composition weighting factor, the masks will be used in computing the composition weights.

In summary, I prefer not to move all these functions.